frontend/controller/serviceworkers/push.js

```diff
@@ -0,0 +1,158 @@
+import { PUBSUB_INSTANCE } from '@controller/instance-keys.js'
+import { makeNotification } from '@model/notifications/nativeNotification.js'
+import sbp from '@sbp/sbp'
+import setupChelonia from '~/frontend/setupChelonia.js'
+import { NOTIFICATION_TYPE, PUBSUB_RECONNECTION_SUCCEEDED, PUSH_SERVER_ACTION_TYPE, REQUEST_TYPE, createMessage } from '~/shared/pubsub.js'
+
+export default (sbp('sbp/selectors/register', {
+  'push/getSubscriptionOptions': (() => {
+    let cachedVapidInformation
+    return () => {
+      if (
+        cachedVapidInformation &&
+        // Cache the VAPID information for one hour. The server public
+        // information should change very infrequently, if it changes at all.
+        (performance.now() - cachedVapidInformation[0]) < 3600e3
+      ) {
+        return cachedVapidInformation[1]
+      }
+
+      const result = new Promise((resolve, reject) => {
+        const handler = ({ data }) => {
+          if (data.type !== PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY) return
+          sbp('okTurtles.events/off', REQUEST_TYPE.PUSH_ACTION, handler)
+          clearTimeout(timeoutId)
+          resolve({
+            userVisibleOnly: true,
+            applicationServerKey: data.data
+          })
+        }
+        const timeoutId = setTimeout(() => {
+          sbp('okTurtles.events/off', REQUEST_TYPE.PUSH_ACTION, handler)
+          reject(new Error('Timed out requesting VAPID key'))
+        }, 10e3)
+
+        const pubsub = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
+        if (!pubsub) reject(new Error('Missing pubsub instance'))
+
+        const readyState = pubsub.socket.readyState
+        if (readyState !== WebSocket.OPEN) {
+          reject(new Error('WebSocket connection is not open'))
+        }
+
+        sbp('okTurtles.events/on', REQUEST_TYPE.PUSH_ACTION, handler)
+        pubsub.socket.send(createMessage(
+          REQUEST_TYPE.PUSH_ACTION,
+          { action: PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY }
+        ))
```

Comment:
The `setTimeout` at the top here is not turned off until a response is received, and yet in between the response and the `setTimeout`, there are at least two immediate opportunities for `reject` to be called before the `timeoutId` is cleared (resulting in an unnecessary second rejection to the already-rejected promise should any of those two `reject` calls get called).

So I would recommend moving the call to the `setTimeout` forward, so that it's done immediately before  line 43.